### PR TITLE
Feature/battery

### PR DESCRIPTION
**Pull request by athackst**

### CODE REVIEW

File: src/task_behavior_ros/batterystate.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import rospy
+
+from sensor_msgs.msg import BatteryState
+
+from task_behavior_engine.tree import NodeStatus
+from task_behavior_ros.topic import TopicMonitor
+
+
+class ChargeCompleteMonitor(TopicMonitor):
+
```
Comment (tinawright) [inline src/task_behavior_ros/batterystate.py:23]: nit: extra newline

File: src/task_behavior_ros/batterystate.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    """
+    Checks if msg indicates charging is complete.
+    @param name [str] The name of this node
+    @param name [topic_name] The name of the topic
```
Comment (tinawright) [inline src/task_behavior_ros/batterystate.py]: i think this should be 
@param topic_name [str] ...

File: src/task_behavior_ros/batterystate.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        @param max_charge [float] Number between (0., 1.) that indicates charging percentage
+    returns NodeStatus.SUCCESS if charging complete
+    returns NodeStatus.ACTIVE if charging
+    returns NodeStatus.FAIL if not charging
+    """
+
+    def __init__(self, name, topic_name, queue_size=1, *args, **kwargs):
+        super(ChargeCompleteMonitor, self).__init__(
+            name=name, topic_name=topic_name, topic_type=BatteryState,
+            cb=self.charge_complete_cb, queue_size=1, *args, **kwargs)
```
Comment (tinawright) [inline src/task_behavior_ros/batterystate.py]: you have queue_size default set twice, here and on the args to this method. I think once is enough.

File: src/task_behavior_ros/batterystate.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+    NodeData:
+        @param max_charge [float] Number between (0., 1.) that indicates charging percentage
+    returns NodeStatus.SUCCESS if charging complete
+    returns NodeStatus.ACTIVE if charging
+    returns NodeStatus.FAIL if not charging
+    """
+
+    def __init__(self, name, topic_name, queue_size=1, *args, **kwargs):
+        super(ChargeCompleteMonitor, self).__init__(
```
Comment (tinawright) [inline src/task_behavior_ros/batterystate.py]: not sure what version of python, but if 3+ can just do "super().__init__("

File: src/task_behavior_ros/batterystate.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+                return NodeStatus(NodeStatus.ACTIVE, "charging")
+        return NodeStatus(NodeStatus.FAIL, "not charging")
+
+
+class ChargeOKMonitor(TopicMonitor):
+
+    """
+    Checks if msg indicates charge is ok.
+    @param name [str] The name of this node
+    @param name [topic_name] The name of the topic
```
Comment (tinawright) [inline src/task_behavior_ros/batterystate.py]: same param question

File: src/task_behavior_ros/batterystate.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    NodeData:
+       @param min_charge [float] Number between (0., 1.) that indicates minimum charging percentage
+    returns NodeStatus.SUCCESS if charge above min
+    returns NodeStatus.FAIL if charge at or below min
+    """
+
+    def __init__(self, name, topic_name, queue_size=1, *args, **kwargs):
+        super(ChargeOKMonitor, self).__init__(
+            name=name, topic_name=topic_name, topic_type=BatteryState,
+            cb=self.charge_ok_cb, queue_size=1, *args, **kwargs)
```
Comment (tinawright) [inline src/task_behavior_ros/batterystate.py]: same default arg question

File: src/task_behavior_ros/batterystate.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+class ChargeCompleteMonitor(TopicMonitor):
+
+    """
+    Checks if msg indicates charging is complete.
+    @param name [str] The name of this node
+    @param name [topic_name] The name of the topic
+    @param queue_size [int] The size of the message queue
+
+    NodeData:
+        @param max_charge [float] Number between (0., 1.) that indicates charging percentage
```
Comment (tinawright) [inline src/task_behavior_ros/batterystate.py]: nit: newline after this line instead of before since it's part of @params?

File: src/task_behavior_ros/batterystate.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    returns NodeStatus.FAIL if charge at or below min
+    """
+
+    def __init__(self, name, topic_name, queue_size=1, *args, **kwargs):
+        super(ChargeOKMonitor, self).__init__(
+            name=name, topic_name=topic_name, topic_type=BatteryState,
+            cb=self.charge_ok_cb, queue_size=1, *args, **kwargs)
+
+    def charge_ok_cb(self, msg, nodedata):
+        if msg.percentage > nodedata.get_data('min_charge', 1.):
```
Comment (tinawright) [inline src/task_behavior_ros/batterystate.py]: so the default min charge is 1? seems high.

File: test/test_batterystate.py [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+        rospy.sleep(0.1)
+        result = self.charge_monitor.tick()
+        assert_equal(result.status, NodeStatus.SUCCESS)
+
+        # Charge at full, but percentage low, still report success
+        msg.percentage = 0.
+        self.msg_pub.publish(msg)
+        rospy.sleep(0.1)  # let callbacks finish
+        result = self.charge_monitor.tick()
+        assert_equal(result.status, NodeStatus.SUCCESS)
```
Comment (tinawright) [inline test/test_batterystate.py]: this test is funny to me, when would be see a full status with a low percentage?

File: test/test_batterystate.py [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+        rospy.sleep(0.1)
+        result = self.charge_monitor.tick()
+        assert_equal(result.status, NodeStatus.FAIL)
+
+        # Unknown but below max, report FAIL
+        msg.percentage = 0.
+        self.msg_pub.publish(msg)
+        rospy.sleep(0.1)  # let callbacks finish
+        result = self.charge_monitor.tick()
+        assert_equal(result.status, NodeStatus.FAIL)
```
Comment (tinawright) [inline test/test_batterystate.py]: I worry a bit about declaring unknown values as success or fail. perhaps add a line to the class doc specifying this behavior so callers can plan accordingly?

File: test/test_batterystate.py [hunk 11]
```diff
@@ -0,0 +1,282 @@
...
+
+    def test_discharging(self):
+        msg = BatteryState()
+        # Discharging at max, report FAIL
+        msg.power_supply_status = BatteryState.POWER_SUPPLY_STATUS_DISCHARGING
+        msg.percentage = 0.
+        self.msg_pub.publish(msg)
+        rospy.sleep(0.1)
+        result = self.charge_monitor.tick()
+        assert_equal(result.status, NodeStatus.FAIL)
```
Comment (tinawright) [inline test/test_batterystate.py]: why is discharging at max a fail? also percentage 0 is max?

File: src/task_behavior_ros/batterystate.py [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+                return NodeStatus(NodeStatus.SUCCESS, "charge above max")
+            else:
+                return NodeStatus(NodeStatus.ACTIVE, "charging")
+        return NodeStatus(NodeStatus.FAIL, "not charging")
+
+
+class ChargeOKMonitor(TopicMonitor):
+
+    """
+    Checks if msg indicates charge is ok.
```
Comment (tinawright) [inline src/task_behavior_ros/batterystate.py:63]: I think perhaps I don't understand exactly what you mean by this. in the tests if the charge is full this returns FAIL. I would expect full charge to return success.

**Review by tinawright — CHANGES_REQUESTED** [review]
nice tests!

File: src/task_behavior_ros/batterystate.py (hunk 1 above)
Comment (athackst) [reply src/task_behavior_ros/batterystate.py:23]: My auto-formatter did this, so I think it's a PEP-8 thing

https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings ?

File: src/task_behavior_ros/batterystate.py (hunk 2 above)
Comment (athackst) [reply src/task_behavior_ros/batterystate.py]: right you are

File: src/task_behavior_ros/batterystate.py (hunk 7 above)
Comment (athackst) [reply src/task_behavior_ros/batterystate.py]: I wanted to separate it out, since you assign it differently.  But +1 to new line after.

File: src/task_behavior_ros/batterystate.py (hunk 4 above)
Comment (athackst) [reply src/task_behavior_ros/batterystate.py]: yeah.. indigo is stuck with 2.7

File: src/task_behavior_ros/batterystate.py (hunk 3 above)
Comment (athackst) [reply src/task_behavior_ros/batterystate.py]: woops,  I think that this makes what you set for queue_size not pass through.  updated

File: src/task_behavior_ros/batterystate.py (hunk 12 above)
Comment (athackst) [reply src/task_behavior_ros/batterystate.py:63]: What if the charge "state" is "FULL" but the charge percentage is super low?  Ignore or go back to base?  I err'd on the side of "say things are wrong" -> ChargeOK = FAIL

File: src/task_behavior_ros/batterystate.py (hunk 5 above)
Comment (athackst) [reply src/task_behavior_ros/batterystate.py]: yep, fixed.

File: src/task_behavior_ros/batterystate.py (hunk 6 above)
Comment (athackst) [reply src/task_behavior_ros/batterystate.py]: yep, removed

File: src/task_behavior_ros/batterystate.py (hunk 8 above)
Comment (athackst) [reply src/task_behavior_ros/batterystate.py]: By default, should we say what a "good" minimal charge is?  Can we know?  I err'd on the side of, if not set, basically always say it's not ok.

We could default to always say it's ok, but then you might not realize you didn't set it and run out of battery.

File: test/test_batterystate.py (hunk 9 above)
Comment (athackst) [reply test/test_batterystate.py]: When a battery state message is being stupid?

I tend to treat unit tests as edge cases.  It doesn't matter if it would never happen, but we should do the Right Thing if it does.

So, to me this was "if the battery isn't reporting a good charge level, but is reporting that the battery is full, should we disconnect from the charger?  -> yes (maybe it's "full" because of temperature/voltage reasons? maybe percentage is all messed up?)

File: test/test_batterystate.py (hunk 11 above)
Comment (athackst) [reply test/test_batterystate.py]: This comment may not be in the right place?

The test is if the power supply is charging and the percentage is below min, FAIL

File: test/test_batterystate.py (hunk 10 above)
Comment (athackst) [reply test/test_batterystate.py]: I think this comment may be in the wrong place?

If you don't fill in the power_supply_status it defaults to "unknown", in that case I would think you would need to rely only on charging percentage.

File: src/task_behavior_ros/batterystate.py [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+        super(ChargeCompleteMonitor, self).__init__(
+            name=name, topic_name=topic_name, topic_type=BatteryState,
+            cb=self.charge_complete_cb, *args, **kwargs)
+
+    def charge_complete_cb(self, msg, nodedata):
+        rospy.loginfo('received msg: ' + str(msg))
+        if msg.power_supply_status == BatteryState.POWER_SUPPLY_STATUS_FULL:
+            return NodeStatus(NodeStatus.SUCCESS, "Power supply state full")
+        if msg.power_supply_status == BatteryState.POWER_SUPPLY_STATUS_CHARGING:
+            if msg.percentage > nodedata.get_data('max_charge', 1.):
```
Comment (LauraStelzner-TRI) [inline src/task_behavior_ros/batterystate.py]: Should this be >= ? Or with how a battery is charged and reported do we actually want it to get greater than 1.0(assuming that is the specified max)?

File: test/test_batterystate.py (hunk 11 above)
Comment (LauraStelzner-TRI) [reply test/test_batterystate.py]: I also don't understand why this case is a fail.

File: test/test_batterystate.py [hunk 14]
```diff
@@ -0,0 +1,282 @@
...
+        self.msg_pub.publish(msg)
+        rospy.sleep(0.1)  # let callbacks finish
+        result = self.charge_monitor.tick()
+        assert_equal(result.status, NodeStatus.FAIL)
+
+    def test_discharging(self):
+        msg = BatteryState()
+        # Discharging at max, report FAIL
+        msg.power_supply_status = BatteryState.POWER_SUPPLY_STATUS_DISCHARGING
+        msg.percentage = 0.
```
Comment (LauraStelzner-TRI) [inline test/test_batterystate.py]: Confused by the comment 'at max', is that referring to min_charge  or to percentage? Why is percentage a 0? Shouldn't it be 1. if its max?

File: test/test_batterystate.py [hunk 15]
```diff
@@ -0,0 +1,282 @@
...
+
+    def test_full(self):
+        msg = BatteryState()
+        # Charge at full, report FAIL.
+        msg.power_supply_status = BatteryState.POWER_SUPPLY_STATUS_FULL
+        msg.percentage = 0.
+        self.msg_pub.publish(msg)
+        rospy.sleep(0.1)
+        result = self.charge_monitor.tick()
+        assert_equal(result.status, NodeStatus.FAIL)
```
Comment (LauraStelzner-TRI) [inline test/test_batterystate.py]: Is the failing for this case because percentage is 0 and the status is full?

File: test/test_batterystate.py [hunk 16]
```diff
@@ -0,0 +1,282 @@
...
+    module_name = sys.modules[__name__].__file__
+
+    parser = argparse.ArgumentParser(description='Perform unit test.')
+    parser.add_argument(
+        '--gtest_output', nargs='?', default='test.xml')
+
+    args, unknown = parser.parse_known_args()
+
+    noseargs = [sys.argv[0], module_name, '--with-xunit',
+                '--xunit-file='+str(args.gtest_output.lstrip('xml:'))]
```
Comment (LauraStelzner-TRI) [inline test/test_batterystate.py:344]: Is the xunit required for Jenkins integration?

File: test/test_topic.py [hunk 17]
```diff
@@ -242,17 +241,52 @@ def test_cb(self):
...
         assert_equal(1, self.blackboard.get('count', self.topic_monitor._id))
 
         self.msg_pub.publish(msg)
-        rospy.sleep(0.5)
+        rospy.sleep(0.1)
         result = self.topic_monitor.tick()
         assert_equal(result.status, NodeStatus.SUCCESS)
         assert_equal(2, self.blackboard.get('count', self.topic_monitor._id))
 
+    def test_latch_cb(self):
```
Comment (LauraStelzner-TRI) [inline test/test_topic.py:256]: Is there a case for testing the value of what is latched?

File: test/test_topic.py [hunk 18]
```diff
@@ -293,14 +327,14 @@ def test_cancel(self):
         msg = String()
         msg.data = "test_string"
         self.msg_pub.publish(msg)
-        rospy.sleep(0.5)
+        rospy.sleep(0.2)
```
Comment (LauraStelzner-TRI) [inline test/test_topic.py:349]: Why .2 instead of .1?

**Review by LauraStelzner-TRI — COMMENTED** [review]
Have some questions to clarify my understanding.  See below.

File: test/test_batterystate.py (hunk 16 above)
Comment (athackst) [reply test/test_batterystate.py:344]: Yep, it's to create the appropriately formatted output file for Jenkins

File: src/task_behavior_ros/batterystate.py (hunk 13 above)
Comment (athackst) [reply src/task_behavior_ros/batterystate.py]: Either is probably valid, but probably should be explicitly stated.  I think you're probably right in that >= makes more semantic sense.

File: test/test_batterystate.py (hunk 15 above)
Comment (athackst) [reply test/test_batterystate.py]: I think something funky is going on with git review.  I'm seeing this on line 178 which is in the charging area

File: test/test_batterystate.py (hunk 14 above)
Comment (athackst) [reply test/test_batterystate.py]: rgr, percentage wrong

File: test/test_topic.py (hunk 17 above)
Comment (athackst) [reply test/test_topic.py:256]: Ah, that would be a good test.  Adding.

File: test/test_batterystate.py [hunk 19]
```diff
@@ -0,0 +1,282 @@
...
+        # Discharging but below min, report FAIL
+        msg.percentage = .1
+        self.msg_pub.publish(msg)
+        rospy.sleep(0.1)  # let callbacks finish
+        result = self.charge_monitor.tick()
+        assert_equal(result.status, NodeStatus.FAIL)
+
+    def test_full(self):
+        msg = BatteryState()
+        # Charge at full, report FAIL.
```
Comment (tinawright) [inline test/test_batterystate.py]: tina confused

**Review by tinawright — APPROVED** [review]

**Review by athackst — COMMENTED** [review]
Changed the behavior so that max_charge and min_charge must be set or they report FAIL.

**Review by LauraStelzner-TRI — APPROVED** [review]
LGTM
